Add authorization logic

nsklikas opened the pull request:
IAM-1559

### Description
Adds app authorization logic. Based on a user's groups we decide whether they should be allowed to access the app they are trying to access or not. The default behavior is to deny access, this means that if the user has no groups or if the app has not been configured then the request is denied.

Most of the authorization logic is copied from the admin UI, I used the latest version of the ofga sdk and had to make some changes compared to the admin UI. Main difference to the admin UI implementation is that the Openfga Tuples are not exposed to outside of the `authorization` package.
The authorization model is copied from the admin UI, the only change is:
```
type client
  relations
    define can_access: [user, group#member] or admin from privileged
...
```
Need to iterate on whether we wish change the api more (we should probably use a different type for the authz api groups)

The groups are written as base64 encodings right now to openfga. This is required as groups may contain characters that are not allowed in openfga (e.g. whitespace)

- Adds authz logic
- Adds ofga integration

### Testing

To test it, add the salesforce credentials to `start.sh`, a github client creds to `.env` and run it.
Go to http://127.0.0.1:4446 and try to log in, this should fail.
Add a tuple with your group to the ofga store (you can find the client-id, store-id and model-id in the run of `start.sh`:
```
fga tuple write --store-id <store-id> --model-id <model-id> group:SWRlbnRpdHk=#member can_access client:<client-id>
```
NOTE: `SWRlbnRpdHk=` is the base64 encoding of the group `Identity`

Trying to login again should succeed, assuming you are trying to login with an email that matches what is in HRc.


### TODO
- Fix the authz model
- Add batch check for audience checks
- Make default behavior configurable

BarcoMasile reviewed (approved):
Amazing work.
I just wonder if we need to replicate the whole authz model for OpenFGA but it doesn't need to be talked about right now.

nsklikas commented on internal/authorization/util.go:
```diff
@@ -0,0 +1,26 @@
+	return "client:" + clientId
+}
+
+func GroupTuple(groupId string) string {
+	// Groups may include invalid characters for an ofga resource (e.g. spaces)
+	// that's why b64 encode them
+	// TODO: revisit this
+	return "group:" + base64.StdEncoding.EncodeToString([]byte(groupId))
+}
```
I agree, will address this in a different PR once database support has been implement it. Will add a clearer TODO message for this, so it does not fall through the crack.

nsklikas commented:
- Removed unnecessary code that was copied from the admin UI (e.g. worker pool)
- Made ofga model minimal
- Add batch check
- Moved authz logic to service layer

wood-push-melon commented on internal/authorization/authorization.go:
```diff
@@ -0,0 +1,109 @@
+	"github.com/canonical/hook-service/internal/tracing"
+)
+
+var ErrInvalidAuthModel = fmt.Errorf("Invalid authorization model schema")
```
just a nitpick:

(follow the go convention) `invalid authorization model schema`

wood-push-melon commented on internal/openfga/client.go:
```diff
@@ -0,0 +1,460 @@
+
+	check, err := c.c.CheckExecute(r)
+	if err != nil {
+		c.logger.Infof("body args: %s %s %s", user, relation, object)
```
a minor question:

Is this for debugging purposes? If so, we could include the information in the next error log

nsklikas commented on internal/openfga/client.go:
```diff
@@ -0,0 +1,460 @@
+	errString = append(errString, "error while performing Check operation:")
+
+	for _, check := range *data.Result {
+		allowed = allowed && *check.Allowed
```
I think there is value in returning all the errors and it shouldn't be a performance issue to iterate over the whole array, I will keep it as it is for now.
